lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -0,0 +1,202 @@
+            in caplog.text
+        )
+
+
+class TestSendDirectory:
+    """Test ToolMeister._send_directory()'s use of ._create_tar()"""
+
+    # Record all the mock functions called by a test
+    functions_called = []
+
+    directory = tmp_dir / f"{tm_params['hostname']}"
+
+    @staticmethod
+    def fake_create_tar(returncode: int, stdout: bytes):
+        def f(directory: pathlib.Path, tar_file: pathlib.Path):
+            return subprocess.CompletedProcess(
+                args=[], returncode=returncode, stdout=stdout, stderr=None
+            )
+
+        __class__.functions_called.append("fake_create_tar")
+        return f
+
+    @responses.activate
+    def test_tar_create_success(self, tool_meister, monkeypatch):
+        """This test should pass the tar creation in send directory"""
+
+        def fake_unlink(*args):
+            assert args[0] == pathlib.Path(f"{self.directory}.tar.xz")
+            self.functions_called.append("fake_unlink")
+
+        def fake_open(*args):
+            assert args[0] == pathlib.Path(f"{self.directory}.tar.xz")
+            self.functions_called.append("fake_open")
+            return io.StringIO()
+
+        def fake_rmtree(directory: pathlib.Path):
+            assert directory == tmp_dir
+            self.functions_called.append("fake_rmtree")
+
+        def fake_md5(tar_file: pathlib.Path):
+            assert tar_file == pathlib.Path(f"{self.directory}.tar.xz")
+            self.functions_called.append("fake_md5")
+            return 10, "random_md5"
+
+        monkeypatch.setattr(shutil, "rmtree", fake_rmtree)
+        monkeypatch.setattr("pbench.agent.tool_meister.md5sum", fake_md5)
+        monkeypatch.setattr(pathlib.Path, "unlink", fake_unlink)
+        monkeypatch.setattr(pathlib.Path, "open", fake_open)
+
+        monkeypatch.setattr(tool_meister, "_create_tar", self.fake_create_tar(0, b""))
+
+        url = (
+            f"http://{tm_params['tds_hostname']}:{tm_params['tds_port']}/uri"
+            f"/ctx/{tm_params['hostname']}"
+        )
+        responses.add(responses.PUT, url, status=HTTPStatus.OK, body="succeeded")
+
+        failures = tool_meister._send_directory(self.directory, "uri", "ctx")
+        functions_called, self.functions_called = self.functions_called, []
+        assert functions_called == [
+            "fake_create_tar",
+            "fake_md5",
+            "fake_open",
+            "fake_rmtree",
+            "fake_unlink",
+        ]
+        assert failures == 0
+
+    def test_tar_create_failure(self, tool_meister, monkeypatch):
+        """Check if the tar creation error is properly captured in send_directory"""
+        monkeypatch.setattr(
+            tool_meister,
+            "_create_tar",
+            self.fake_create_tar(1, b"Error in tarball creation"),
+        )
+
+        with pytest.raises(ToolMeisterError) as exc:
+            failures = tool_meister._send_directory(self.directory, "uri", "ctx")
+            assert failures == 1
```

Comment:
Are you sure you want this `assert` statement inside the `with` block?

Reply:
Good catch, that is bad.

Reply:
I don't know that it is bad, per se -- if the assertion fails, the test will certainly fail -- but it's probably not good.  🙂

Reply:
That assertion never fires so it was removed.

Reply:
Oh, yes...of course.  (I.e., "Well, duh!"  😆)